lib/handler/connect.c

```diff
@@ -123,6 +124,13 @@ struct st_connect_generator_t {
 
 static void record_error(struct st_connect_generator_t *self, const char *error_type, const char *details, const char *rcode)
 {
+    if (H2O_CONN_IS_PROBED(CONNECT_ERROR, self->src_req->conn)) {
+        char msg1[64], msg2[64]; // STR_LEN in src/h2olog/misc/gen_raw_tracer.py
```

Comment:
IIRC h2olog discards a string that is longer than STR_LEN, so we don't need to adjust it here.

Also, as you might know, the BPF program in h2olog does not compile if an event object is too large so STR_LEN is restricted to a small value, but in this case, it seems that we don't need to contact error_type and rcode. 

So 
`H2O_PROBE_CONN(CONNECT_ERROR, self->src_req->conn, self->src_req->req_index, error_type, rcode, details);` might be enough. What do you think?

Reply:
Ah, it's the case where the event object is too long!:

> error: /virtual/main.c:1806:13: in function trace_quicly__stream_send i32 (%struct.pt_regs*): Looks like the BPF stack limit of 512 bytes is exceeded. Please move large on stack variables into BPF per-cpu array map.

Please ignore the above comment.

Reply:
Yeah I think it might be possible to reduce stack pressure pretty easily.

At the moment, `trace_quicly__stream_send` looks like this:
```c
// quicly:stream_send
int trace_quicly__stream_send(struct pt_regs *ctx) {
  const void *buf = NULL;
  struct h2olog_event_t event = { .id = H2OLOG_EVENT_ID_QUICLY_STREAM_SEND, .tid = (uint32_t)bpf_get_current_pid_tgid(), };

  // struct st_quicly_conn_t * conn
  uint8_t conn[sizeof_st_quicly_conn_t] = {};
  bpf_usdt_readarg(1, ctx, &buf);
  bpf_probe_read(&conn, sizeof_st_quicly_conn_t, buf);
  event.stream_send.conn_master_id = get_st_quicly_conn_t__master_id(conn);
  // int64_t at
  bpf_usdt_readarg(2, ctx, &event.stream_send.at);
  // struct st_quicly_stream_t * stream
  uint8_t stream[sizeof_st_quicly_stream_t] = {};
  bpf_usdt_readarg(3, ctx, &buf);
  bpf_probe_read(&stream, sizeof_st_quicly_stream_t, buf);
  event.stream_send.stream_stream_id = get_st_quicly_stream_t__stream_id(stream);
  // uint64_t off
  bpf_usdt_readarg(4, ctx, &event.stream_send.off);
  // size_t len
  bpf_usdt_readarg(5, ctx, &event.stream_send.len);
  // int is_fin
  bpf_usdt_readarg(6, ctx, &event.stream_send.is_fin);

  if (events.perf_submit(ctx, &event, sizeof(event)) != 0)
    bpf_trace_printk("failed to perf_submit in trace_quicly__stream_send\n");

  return 0;
}
```

The pressure comes from `uint8_t conn[sizeof_st_quicly_conn_t]` and `uint8_t stream[sizeof_st_quicly_stream_t]` sharing the same stack.

I think (or would hope) that if we could use blocks (`{}`) for each argument so that `conn` and `stream` do not coexist at one time, then we can free 100 bytes of stack space.

Reply:
> I think (or would hope) that if we could use blocks ({}) for each argument so that conn and stream do not coexist at one time, then we can free 100 bytes of stack space.


Confirmed https://github.com/h2o/h2o/pull/3051 makes this PR compile with three string params. Let's remove the hack to merge strings.

Reply:
Done in b5c36230f.

Reply:
Done in d489a36.